backend/src/test/java/ee/tlu/forum/controller/PostControllerTest.java

```diff
@@ -0,0 +1,244 @@
+    void getPosts() throws Exception {
+        // given
+        User user = new User(1L,
+                "user1",
+                "test1@test.com",
+                "aaa",
+                new ArrayList<>(),
+                "",
+                "",
+                new ArrayList<>(),
+                new ArrayList<>());
+        Thread t1 = new Thread(1L, "content", "title", user, new ArrayList<>());
+        Post p1 = new Post(1L, "text", user, t1);
+
+        when (postService.getAllPosts())
+                .thenReturn(Arrays.asList(p1));
+
+        // when then
+        mockMvc.perform(MockMvcRequestBuilders.get("/api/posts"))
+                .andDo(MockMvcResultHandlers.print())
+                .andExpect(MockMvcResultMatchers.status().is(200))
+                .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.size()", Matchers.is(1)))
+                .andExpect(MockMvcResultMatchers.jsonPath("$[0].id", Matchers.is(1)))
+                .andExpect(MockMvcResultMatchers.jsonPath("$[0].text", Matchers.is("text")));
+    }
+
+    @Test
+    @DisplayName("Returns a post by ID at - GET /api/thread/{id}")
+    void getPostById() throws Exception {
+        // given
+        User user = new User(1L,
+                "user1",
+                "test1@test.com",
+                "aaa",
+                new ArrayList<>(),
+                "",
+                "",
+                new ArrayList<>(),
+                new ArrayList<>());
+        Thread t1 = new Thread(1L, "content", "title", user, new ArrayList<>());
+        Post p1 = new Post(1L, "text", user, t1);
+
+        when (postService.getPostById(any()))
+                .thenReturn(p1);
+
+        // when then
+        mockMvc.perform(MockMvcRequestBuilders.get("/api/post/1"))
+                .andDo(MockMvcResultHandlers.print())
+                .andExpect(MockMvcResultMatchers.status().is(200))
+                .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.id", Matchers.is(1)))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.text", Matchers.is("text")));
+    }
+
+    @Test
+    @DisplayName("Returns created post - POST /api/post/add")
+    void createPost() throws Exception {
+        // given
+        User user = new User(1L,
+                "user1",
+                "test1@test.com",
+                "aaa",
+                new ArrayList<>(),
+                "",
+                "",
+                new ArrayList<>(),
+                new ArrayList<>());
+        Thread t1 = new Thread(1L, "content", "title", user, new ArrayList<>());
+        Post p1 = new Post(1L, "text", user, t1);
+
+        AddNewPostInput form = new AddNewPostInput();
+        form.setUsername("user1");
+        form.setText("text");
```

Comment:
If I get this right, this is a new post, in addition to p1? If that's the case, then I'd change the text of this new post so the two do not match and the test below actually differentiates between them 🙂

Reply:
P1 is the post that will be returned - not yet created in the context of the repository. We construct the output of the service layer and then go to the endpoint of the controller, which then calls that service and returns the constructed output (eg a new post, p1) and the controller has to get the same output.

Another way is to test with the H2 database and also mock the repository, but that has its own plethora of complexity!

Reply:
Ah OK, I'd probably mock out both states (original and then edited) and check that it actually changes but that's probably just me.

Reply:
Ohhh, I misunderstood the code then. My bad.